src/realizations/catchment/Bmi_C_Formulation.cpp

```diff
@@ -0,0 +1,161 @@
+
+typedef std::vector<std::string> vector_str;
+
+Bmi_C_Formulation::Bmi_C_Formulation(std::string id, forcing_params forcing_config, utils::StreamHandler output_stream)
+    : Bmi_Formulation<models::bmi::Bmi_C_Adapter>(id, forcing_config, output_stream) { }
+
+std::string Bmi_C_Formulation::get_formulation_type() {
+    return "bmi_c";
+}
+
+void Bmi_C_Formulation::create_formulation(boost::property_tree::ptree &config, geojson::PropertyMap *global) {
+    geojson::PropertyMap options = this->interpret_parameters(config, global);
+    inner_create_formulation(options, false);
+}
+
+void Bmi_C_Formulation::create_formulation(geojson::PropertyMap properties) {
+    inner_create_formulation(properties, true);
+}
+
+std::string Bmi_C_Formulation::get_output_header_line(std::string delimiter) {
+    return boost::algorithm::join(get_output_header_fields(), delimiter);
+}
+
+std::string Bmi_C_Formulation::get_output_line_for_timestep(int timestep, std::string delimiter) {
+    // Check if the timestep is in bounds for the fluxes vector, and handle case when it isn't
+    if (timestep > get_bmi_model()->get_last_processed_time_step()) {
+        return "";
+    }
+    std::string output_str;
+
+    for (const std::string& name : get_output_variable_names()) {
+        output_str += (output_str.empty() ? "" : ",") + std::to_string(get_var_value_as_double(timestep, name));
+    }
+    return output_str;
+}
+
+/**
+ * Get the model response for this time step.
+ *
+ * Get the model response for this time step, execute the backing model formulation one or more times if the time step
+ * of the given index has not already been processed.
+ *
+ * Function assumes the backing model has been fully initialized an that any additional input values have been applied.
+ *
+ * The function will return the value of the primary output variable (see `get_bmi_main_output_var()`) for the given
+ * time step. The type returned will always be a `double`, with other numeric types being cast if necessary.
+ *
+ * Because of the nature of BMI, the `t_delta` parameter is ignored, as this cannot be passed meaningfully via
+ * the `update()` BMI function.
+ *
+ * @param t_index The index of the time step for which to run model calculations.
+ * @param d_delta_s The duration, in seconds, of the time step for which to run model calculations.
+ * @return The total discharge of the model for the given time step.
+ */
+double Bmi_C_Formulation::get_response(time_step_t t_index, time_step_t t_delta) {
+    if (get_bmi_model() == nullptr) {
+        throw std::runtime_error("Trying to process response of improperly created BMI C formulation.");
+    }
+    if (t_index < 0) {
+        throw std::invalid_argument("Getting response of negative time step in BMI C formulation is not allowed.");
+    }
+
+    int last_processed_time_step = get_bmi_model()->get_last_processed_time_step();
+    while (last_processed_time_step++ < t_index) {
+        get_bmi_model()->Update();
```

Comment:
Is this significant, and if it is, is it addressable?  

A call to the BMI `Update` function can only advance the model by another time step, whatever the "next" time step is.  I don't see a clearly viable way to meaningfully pass the `t_delta` value to the model.  So, when asking for the response for time step `t_index`, what else could be done other than processing one time step at a time until the `t_index`-th time step has been processed?

It occurs to me as I'm writing this that `Update_until` could potentially be of some help depending on the way it is implemented.  I'd originally thought of it updating a series of time steps until the model's `current_time` passed the given point.  But maybe it advances by a _single_ time step of whatever size was needed to advance to the _until_ time?  I'm not sure which would be "correct" (or if either/neither could be), and it may not matter if the particulars vary from model to model.